src/test/java/com/google/api/codegen/testdata/php/php_library.baseline

```diff
@@ -2728,7 +2319,323 @@ class LibraryServiceClient extends LibraryServiceGapicClient
+                    'metadataReturnType' => '\Google\Example\Library\V1\GetBigBookMetadata',
+                ]
+            ],
+            'ListShelves' => [
+                'pageStreaming' => [
+                    'requestPageTokenGetMethod' => 'getPageToken',
+                    'requestPageTokenSetMethod' => 'setPageToken',
+                    'responsePageTokenGetMethod' => 'getNextPageToken',
+                    'resourcesGetMethod' => 'getShelves'
+                ]
+            ],
+            'ListBooks' => [
+                'pageStreaming' => [
+                    'requestPageTokenGetMethod' => 'getPageToken',
+                    'requestPageTokenSetMethod' => 'setPageToken',
+                    'requestPageSizeGetMethod' => 'getPageSize',
+                    'requestPageSizeSetMethod' => 'setPageSize',
+                    'responsePageTokenGetMethod' => 'getNextPageToken',
+                    'resourcesGetMethod' => 'getBooks'
+                ]
+            ],
+            'ListStrings' => [
+                'pageStreaming' => [
+                    'requestPageTokenGetMethod' => 'getPageToken',
+                    'requestPageTokenSetMethod' => 'setPageToken',
+                    'requestPageSizeGetMethod' => 'getPageSize',
+                    'requestPageSizeSetMethod' => 'setPageSize',
+                    'responsePageTokenGetMethod' => 'getNextPageToken',
+                    'resourcesGetMethod' => 'getStrings'
+                ]
+            ],
+            'FindRelatedBooks' => [
+                'pageStreaming' => [
+                    'requestPageTokenGetMethod' => 'getPageToken',
+                    'requestPageTokenSetMethod' => 'setPageToken',
+                    'requestPageSizeGetMethod' => 'getPageSize',
+                    'requestPageSizeSetMethod' => 'setPageSize',
+                    'responsePageTokenGetMethod' => 'getNextPageToken',
+                    'resourcesGetMethod' => 'getNames'
+                ]
+            ],
+            'StreamShelves' => [
+                'grpcStreaming' => [
+                    'grpcStreamingType' => 'ServerStreaming',
+                    'resourcesGetMethod' => 'getShelves',
+                ]
+            ],
+            'StreamBooks' => [
+                'grpcStreaming' => [
+                    'grpcStreamingType' => 'ServerStreaming',
+                ]
+            ],
+            'DiscussBook' => [
+                'grpcStreaming' => [
+                    'grpcStreamingType' => 'BidiStreaming',
+                ]
+            ],
+            'MonologAboutBook' => [
+                'grpcStreaming' => [
+                    'grpcStreamingType' => 'ClientStreaming',
+                ]
+            ],
+        ]
+    ]
+];
+============== file: src/Example/Library/V1/resources/library_service_rest_client_config.php ==============
+<?php
+
+return [
+    'interfaces' => [
+        'google.example.library.v1.Labeler' => [
```

Comment:
A question about this approach - why do we need to place methods that use a rerouteToGrpcInterface setting in a different array? Can we not follow the approach of the settings json file and put them under the main interface (see https://github.com/googleapis/toolkit/pull/1787/files#diff-36f32ff1b2217a381ca1eaeeea45dfe6R2298) an then have a rerouteToInterface setting in the array itself? It seems like this would be consistent, and would also simplify the java implementation.

Reply:
The interface name gets passed into the `Call` class [here](https://github.com/googleapis/gax-php/blob/json-transport/src/ApiCore/GapicClientTrait.php#L235). We use that combination of the interface name/method name to register the correct gRPC call or to perform a lookup for REST configuration data.

In order to achieve what you mention for the case of REST, I believe we would need to have transport specific logic living in the trait to pass in the main interface name instead. If possible, I would like to keep transport specific logic out of the trait.

Is there a purpose to the "interfaces" key in the .json config if we end up putting everything under the same blanket?

Reply:
Thought on this a little more, WDYT about modifying the `Call` class to look like the following?

```php
$call = new Call(
    $interface,
    $method,
    $decodeType,
    $message,
    $descriptor,
    $callType,
    $rerouteToInterface
);
```

Then in the transport implementations, if `$rerouteToInterface` is set use that over the `$interface` when constructing the request?

Reply:
Hmm, but why is this transport dependent? Both gRPC and REST need to support it. How does grpc do it now?

Reply:
I don't understand why having multiple interfaces is a bad idea. Isn't the call actually made to a separate service URL, so it makes sense? I feel like `rerouteToInterface` may be a gRPC-specific idea, or at least I'm confused by it.

Reply:
> Hmm, but why is this transport dependent? Both gRPC and REST need to support it. How does grpc do it now?

gRPC needs the `rerouteToInterface` when triggering [`\Grpc\BaseStub::_simpleRequest`](https://github.com/googleapis/gax-php/blob/json-transport/src/ApiCore/Transport/GrpcTransport.php#L146), but REST only really needs it in order to look up the correct URI template in the configuration file.

My original concern was if we move everything under one interface, we would need to do something like the following in the trait:

```php
    if ($transportType === 'REST') {
        $interfaceName = $this->serviceName;
    }

    $call = new Call(
        $this->buildMethod($interfaceName, $methodName),
        $decodeType,
        $request,
        $descriptor,
        $callType
    );
```

But this was superseded by the suggestion which followed.

>I don't understand why having multiple interfaces is a bad idea. Isn't the call actually made to a separate service URL, so it makes sense?

I side with this. I am having trouble understanding why currently everything lives under one interface (in the config.json, for example), even if they technically may be defined under another.

Reply:
I don't think having multiple interfaces is a bad idea, but I suppose I didn't understand why it was necessary for REST when it isn't (or doesn't seem to be) necessary for grpc.

It doesn't call a separate service URL. In fact in the REST case, I think the final result is completely unchanged. We only use the interface name as a key into the settings: https://github.com/googleapis/gax-php/blob/json-transport/src/ApiCore/RequestBuilder.php#L74

In the grpc case, we need it to make the _simpleRequest call. E.g. for IAM we are replicating this call: https://github.com/googleapis/gax-php/blob/json-transport/src/Cloud/Iam/V1/IAMPolicyGrpcClient.php#L68

I didn't mean to say the current approach is definitely wrong, only that it seemed different and slightly more complex than the grpc case and I didn't understand why.

Reply:
Sorry, my last comment did not see yours @dwsupplee

Reply:
No worries!

I think if we do keep multiple interfaces, it should be under the condition we eventually do the same for the descriptor and json config. If that seems like something which is off the table, does https://github.com/googleapis/toolkit/pull/1787#discussion_r162453900 seem reasonable?

Reply:
I think that comment looks reasonable, especially splitting the interface and method name. Rather than a separate parameter for rerouteToInterface, what about having that as part of the descriptor array?

Reply:
That sounds like a nice idea. For clarification on my end, what do we qualify a descriptor as? Is it a more concrete concept in toolkit (with batchDescriptors, pageDescriptors, etc), or can we use it more generically for "anything which describes this call?"

Reply:
It is more generic. It should be (I believe) features of a call that are fixed/permanent (and perhaps overridable by the user, but perhaps not). In the gapic yaml config, the rerouteToGrpcInterface setting (see here: https://github.com/googleapis/googleapis/blob/master/google/pubsub/v1/pubsub_gapic.yaml#L601) is the same sort of setting as page_streaming settings: https://github.com/googleapis/googleapis/blob/master/google/pubsub/v1/pubsub_gapic.yaml#L560